README.md

```diff
@@ -1,8 +1,380 @@
+
+The builtin type of middleware module is `default` and the builtin service type is `gateway` (`default` is probably a better name, up to change in the future). As you maybe already noticed, the types are somewhat arbitrary as they are not necessarily abstractions but indications about how Zanzibar should treat the modules.
+
+Note that the module classes and types mentioned above are builtin in the sense that Zanzibar knows how to treat a module with given builtin classes and types. We can always easily extend by defining arbitrary module classs and types as long as we register appropriate module generators to the module system.
+
+#### Dependency
+Module dependencies describe the relationships among various modules. The dependency relationship is critical to correctly assemble the modules to a full application.
+
+##### Dependency Injection
+A module is expected to define its immediate or direct dependencies. Zanzibar generates a moudle constructor with depedent modules as parameters, and passes the dependencies to the constructor during initilizaiton.
+
+##### Module Initialization
+Zanzibar also constructs a full global dependency graph for a given set of modules. This graph is used to initialize modules in the correct order, e.g. leaf modules are initialized first and then passed to the constructors of parent modules for initialization.
+
+##### Dependency Rules
+To establish and enforce abstraction boundaries, dependency rules at `ModuleClass` level are necessary. Zanzibar predefines the following dependency rules for built module classes:
+
+ModuleClass | DependsOn | DependedBy
+----------- | --------- | ----------
+client | N/A | middleware, endpoint
+middleware | client | endpoint
+endpoint | client, middleware | service
+servcie | endpoint | N/A
+
+This table exhausts the possible immediate or direct dependency relationships among builtin module classes. Take endpoint module class for example, an endpoint module can depend on client or middleware modules but not endpoint or servcie modules. The resoning for such rules aligns with the abstractions of the module classes represent.
+
+The `ModuleClass` struct has `DependsOn` and `DependedBy` public fields, which makes it simple to extend the dependency rules with custom module class, e.g., we can define a custom module class `task` that abstracts common business workflow by setting its `DependsOn` field to client and `DependedBy` field to endpoint.
+
+### Config
+Configurations are the interface that developers interact with when using the Zanzibar framework, they make up most of Zazibar's API. Various configurarions contain essential meta information of a Zanziabr application and its components. They are source of truth of the application.
+
+**Note**: Currently configurations are in JSON, we plan to migrate to YAML.
+
+#### Config Layout
```

Comment:
You should move this to the beginning since this informs overall the project structure of a zanzibar application, rather than specific to this section (configuration only)

Reply:
The thing with that arrangement is that the content of this section assumes the readers know what a module is.